From 238a05cfc727866fec9e95a9640d85f760b82678 Mon Sep 17 00:00:00 2001
Date: Tue, 19 Feb 2013 12:08:27 +0000
Subject: [PATCH] xvba_fix_pix_fmt_vda_vld_assert

---
 libavutil/utils.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavutil/utils.c b/libavutil/utils.c
index fbfbc49..56a83a7 100644
--- a/libavutil/utils.c
+++ b/libavutil/utils.c
@@ -28,7 +28,7 @@
 
 unsigned avutil_version(void)
 {
-    av_assert0(AV_PIX_FMT_VDA_VLD == 81); //check if the pix fmt enum has not had anything inserted or removed by mistake
+    av_assert0(AV_PIX_FMT_VDA_VLD == 82); //check if the pix fmt enum has not had anything inserted or removed by mistake
     av_assert0(AV_SAMPLE_FMT_DBLP == 9);
     av_assert0(AVMEDIA_TYPE_ATTACHMENT == 4);
     av_assert0(AV_PICTURE_TYPE_BI == 7);
-- 
1.7.10.4

