From e445bbce8e6f090b84a3a7556aefb1168452e560 Mon Sep 17 00:00:00 2001
From: vicbitter <xbmcvb@gmail.com>
Date: Mon, 8 Apr 2013 22:54:19 +0100
Subject: [PATCH] Fix XVBA to use CDisplaySettings::Get().GetPixelRatio()

---
 xbmc/cores/VideoRenderers/LinuxRendererGL.cpp |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/cores/VideoRenderers/LinuxRendererGL.cpp b/xbmc/cores/VideoRenderers/LinuxRendererGL.cpp
index b69de25..4c410f1 100644
--- a/xbmc/cores/VideoRenderers/LinuxRendererGL.cpp
+++ b/xbmc/cores/VideoRenderers/LinuxRendererGL.cpp
@@ -1807,7 +1807,7 @@ void CLinuxRendererGL::RenderXVBA(int index, int field)
     if (g_application.m_pPlayer && g_application.m_pPlayer->IsInMenu())
       m_pVideoFilterShader->SetNonLinStretch(1.0);
     else
-      m_pVideoFilterShader->SetNonLinStretch(pow(g_settings.m_fPixelRatio, g_advancedSettings.m_videoNonLinStretchRatio));
+      m_pVideoFilterShader->SetNonLinStretch(pow(CDisplaySettings::Get().GetPixelRatio(), g_advancedSettings.m_videoNonLinStretchRatio));
 
     m_pVideoFilterShader->Enable();
   }
-- 
1.7.10.4

