From e79403fcf05341c4b79aad4d7ff22c36a7b30466 Mon Sep 17 00:00:00 2001
Date: Sat, 16 Mar 2013 08:57:00 +0000
Subject: [PATCH] xbmc

---
 libavcodec/dxva2_mpeg2.c |   14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/libavcodec/dxva2_mpeg2.c b/libavcodec/dxva2_mpeg2.c
index 7c292bf..938540c 100644
--- a/libavcodec/dxva2_mpeg2.c
+++ b/libavcodec/dxva2_mpeg2.c
@@ -22,12 +22,12 @@
 
 #include "dxva2_internal.h"
 
-#define MAX_SLICES (SLICE_MAX_START_CODE - SLICE_MIN_START_CODE + 1)
 struct dxva2_picture_context {
     DXVA_PictureParameters pp;
     DXVA_QmatrixData       qm;
     unsigned               slice_count;
-    DXVA_SliceInfo         slice[MAX_SLICES];
+    DXVA_SliceInfo         *slice;
+    unsigned int           slice_alloc;
 
     const uint8_t          *bitstream;
     unsigned               bitstream_size;
@@ -220,6 +220,8 @@ static int dxva2_mpeg2_start_frame(AVCodecContext *avctx,
     fill_quantization_matrices(avctx, ctx, s, &ctx_pic->qm);
 
     ctx_pic->slice_count    = 0;
+    ctx_pic->slice          = NULL;
+    ctx_pic->slice_alloc    = 0;
     ctx_pic->bitstream_size = 0;
     ctx_pic->bitstream      = NULL;
     return 0;
@@ -232,9 +234,14 @@ static int dxva2_mpeg2_decode_slice(AVCodecContext *avctx,
     struct dxva2_picture_context *ctx_pic =
         s->current_picture_ptr->f.hwaccel_picture_private;
     unsigned position;
+    DXVA_SliceInfo* slice;
 
-    if (ctx_pic->slice_count >= MAX_SLICES)
+    slice = av_fast_realloc(ctx_pic->slice,
+                            &ctx_pic->slice_alloc,
+                            (ctx_pic->slice_count + 1) * sizeof(DXVA_SliceInfo));
+    if (!slice)
         return -1;
+    ctx_pic->slice = slice;
 
     if (!ctx_pic->bitstream)
         ctx_pic->bitstream = buffer;
@@ -261,6 +268,7 @@ static int dxva2_mpeg2_end_frame(AVCodecContext *avctx)
                                     commit_bitstream_and_slice_buffer);
     if (!ret)
         ff_mpeg_draw_horiz_band(s, 0, avctx->height);
+    av_freep(ctx_pic->slice);
     return ret;
 }
 
-- 
1.7.10.4

